go/vt/vttablet/tabletserver/tx_engine.go

```diff
@@ -149,38 +145,74 @@ func (te *TxEngine) transition(state txEngineState) {
+		if !te.preparedPool.IsOpen() {
+			// We need to redo prepared transactions here to handle vttablet restarts.
+			// If MySQL continues to work fine, then we won't end up redoing the prepared transactions as part of any RPC call
+			// since VTOrc won't call `UndoDemotePrimary`. We need to do them as part of this transition.
+			_ = te.redoPreparedTransactionsLocked()
+		}
+		te.startTransactionWatcher()
+	}
 	te.txPool.Open(te.env.Config().DB.AppWithDB(), te.env.Config().DB.DbaWithDB(), te.env.Config().DB.AppDebugWithDB())
+}
 
-	if te.twopcEnabled && te.state == AcceptingReadAndWrite {
-		// Set the preparedPool to start accepting connections.
-		te.preparedPool.shutdown = false
-		// If there are errors, we choose to raise an alert and
-		// continue anyway. Serving traffic is considered more important
-		// than blocking everything for the sake of a few transactions.
-		// We do this async; so we do not end up blocking writes on
-		// failover for our setup tasks if using semi-sync replication.
-		te.twoPCReady.Add(1)
-		go func() {
-			defer te.twoPCReady.Done()
-			if err := te.twoPC.Open(te.env.Config().DB); err != nil {
-				te.env.Stats().InternalErrors.Add("TwopcOpen", 1)
-				log.Errorf("Could not open TwoPC engine: %v", err)
-			}
-			if err := te.prepareFromRedo(); err != nil {
-				te.env.Stats().InternalErrors.Add("TwopcResurrection", 1)
-				log.Errorf("Could not prepare transactions: %v", err)
-			}
-			te.startTransactionWatcher()
-		}()
+// RedoPreparedTransactions acquires the state lock and calls redoPreparedTransactionsLocked.
+func (te *TxEngine) RedoPreparedTransactions() error {
+	if !te.twopcEnabled {
+		return nil
+	}
+	te.stateLock.Lock()
+	defer te.stateLock.Unlock()
+	return te.redoPreparedTransactionsLocked()
+}
+
+// redoPreparedTransactionsLocked redoes the prepared transactions.
+// If there are errors, we choose to raise an alert and
+// continue anyway. Serving traffic is considered more important
+// than blocking everything for the sake of a few transactions.
+// We do this async; so we do not end up blocking writes on
+// failover for our setup tasks if using semi-sync replication.
+func (te *TxEngine) redoPreparedTransactionsLocked() error {
+	oldState := te.state
+	// We shutdown to ensure no other writes are in progress.
+	te.shutdownLocked()
+	defer func() {
+		te.state = oldState
+	}()
+
+	if err := te.twoPC.Open(te.env.Config().DB); err != nil {
+		te.env.Stats().InternalErrors.Add("TwopcOpen", 1)
+		log.Errorf("Could not open TwoPC engine: %v", err)
+		return err
 	}
+
+	// We should only open the prepared pool and the transaction pool if the opening of twoPC pool is successful.
+	// We use the prepared pool being open to know if we need to redo the prepared transactions.
+	// So if we open the prepared pool and then opening of twoPC fails, we will never end up opening the twoPC pool at all!
+	// This is why opening prepared pool after the twoPC pool is crucial for correctness.
```

Comment:
Do you mean opening the transaction pool after the two PC pool?

Reply:
The code and comment does not match or at least is unclear

Reply:
We're opening the twoPC pool before we open the prepared pool or txpool. The comment explains why that is important

Reply:
If we did it the other way around, and after the prepared pool was opened, and then opening of twoPC fails, we won't open it ever. So we have to open twoPC first.